bittide-shake/data/tcl/HardwareTest.tcl

```diff
@@ -474,22 +520,79 @@ proc print_test_results {done success start_time end_time} {
     }
 }
 
-# Gets the short names of probes which contain 'probe_test_start'
-proc get_test_names {} {
+# Get the test names from the test config file.
+# The test names are the keys of the tests dictionary in the yaml file, exluding
+# the default key, which is used for default values.
+proc get_test_names {test_config} {
     global vio_prefix
-    set start_probes [get_hw_probes ${vio_prefix}/probe_test_start*]
-    if {[expr [llength $start_probes] == 0]} {
-        puts "No probes found with name '*probe_test_start*', which are needed to start tests"
-        exit 1
+    set tests [dict get $test_config tests]
+    set test_names [dict keys $tests]
+    set test_names [lsearch -all -inline -not -exact $test_names default]
+    return $test_names
+}
+
+# Receives the test config, the index of the currently active FPGA and current test name.
+# It sets the extra probes defined in the test config for the specified test and FPGA.
+proc set_extra_probes {yaml_dict fpga_index test_name} {
+    puts -nonewline "Setting extra probes for test: $test_name, fpga: $fpga_index: "
+    global fpga_ids
+    global vio_prefix
+    set default_dict [dict get $yaml_dict default]
+
+    set probe_dicts []
+
+    # Add the default probes to the list of probe_dicts
+    if {[dict exists $default_dict probes]} {
+        lappend probe_dicts [dict get $default_dict probes]
     }
-    set start_probe_names {}
-    foreach start_probe $start_probes {
-        lappend start_probe_names [get_property name.short $start_probe]
+
+    # Add test specific probes
+    set test_dict [dict get $yaml_dict tests $test_name]
+    if {[dict exists $test_dict probes]} {
+        lappend probe_dicts [dict get $test_dict probes]
+    }
+
+    # Add FPGA specific probes
+    if {[dict exists $test_dict targets]} {
+        set target_list [dict get $test_dict targets]
+        foreach target $target_list {
+            if {[dict get $target target index] == $fpga_index} {
+                if {[dict exists $target probes]} {
+                    lappend probe_dicts [dict get $target probes]
+                }
+            }
+        }
+    }
+
+    # For each probe dictionary, set the probes
+    set changed_vios []
+    foreach probe_dict $probe_dicts {
+        dict for {vio_name vio_value} $probe_dict {
+            set probe [get_hw_probes ${vio_prefix}/${vio_name}]
+            if {[lsearch -exact $changed_vios $probe] }  {
+                lappend changed_vios $probe
+            }
+            set bit_width [get_property width $probe]
+            set hex_width [expr {(3 + $bit_width)/4}]
+            set vio_value [format "%0${hex_width}llX" [expr $vio_value]]
```

Comment:
```suggestion
            set vio_value [format "%0${hex_width}llX" [parse_integer $vio_value]]
```

Could you make sure that value that's being passed in corresponds to one of these regexes:

```
0x[0-9A-F]+
0b[0-1]+
0o[0-7]+
[0-9]+
```

`expr` can do arbitrary function calls.

Reply:
Some ChatGPT magic:

```tcl
proc matchNumericFormat {str} {
    # Combine the patterns with logical OR
    set pattern {^(0x[0-9A-Fa-f]+|0b[0-1]+|0o[0-7]+|[0-9]+)$}

    # Check if the string matches any of the patterns
    if {[regexp $pattern $str]} {
        return 1  ;# True, the string matches one of the patterns
    } else {
        return 0  ;# False, the string does not match any of the patterns
    }
}

# Example usage
set strings {0x1A3F 0b101001 0o755 12345 notANumber}

foreach str $strings {
    if {[matchNumericFormat $str]} {
        puts "$str matches one of the patterns."
    } else {
        puts "$str does not match any of the patterns."
    }
}
```

Reply:
And yes you can be sure to get review comments if you copy paste that at verbatim :-)

Reply:
I think you just want to check with `string is entier $vio_value`.

Reply:
Boo, Vivado's Tcl is too old for that. use `wideinteger`.
```tcl
if {![string is wideinteger $vio_value]} {
  puts "\$vio_value not an integer: $vio_value"
  exit 1
}
```
(I don't know why you don't use `error`? Is the result not pleasing to the eye? But I just followed what was there already in the file, which is `puts` / `exit 1`)

Reply:
Finally, this seems to work just fine
```tcl
format %0${hex_width}llX $vio_value
```
(I also removed superfluous quoting)

Reply:
> I don't know why you don't use error? 

I just copied straight from ChatGPT. Like I said

> And yes you can be sure to get review comments if you copy paste that at verbatim :-)

:-)

Reply:
> Finally, this seems to work just fine

Ah of course. Everything is a string in TCL after all!

Reply:
>> I don't know why you don't use `error`?
>
> I just copied straight from ChatGPT. Like I said

Wrong "you" :-). This script, `bittide-shake/data/tcl/HardwareTest.tcl`, uses `puts` and `exit 1` whenever it handles an error condition, so I followed suit in my suggestion, but asked why it was done in this way. This suddenly becomes more relevant, because if you just do `format %0${hex_width}llX $vio_value` in the code and `$vio_value` is not a number, it'll behave pretty much like `error`, and not like `puts` / `exit 1`. So if it was a conscious decision to use `puts` and `exit 1` for handling all error conditions, this will have to be handled explicitly. If normal Tcl error handling is acceptable, you can just not check `$vioValue`, invoke `format` and let that throw an error.

Reply:
At a glance, the main difference between `puts` / `exit 1` and Tcl error handling is that the latter outputs a stack trace rather than just the message. They both exit with a return code 1 in Vivado, and both output their messsages on `stdout`.

Reply:
Then I guess `error` is preferable!